README.md
# ieee754-buffer
Copyright (c) 2018-2019 Rafael da Silva Rocha.  
Copyright (c) 2013 DeNA Co., Ltd.  
Copyright (c) 2010, Linden Research, Inc  
https://github.com/rochars/ieee754-buffer

[![NPM version](https://img.shields.io/npm/v/ieee754-buffer.svg?style=for-the-badge)](https://www.npmjs.com/package/ieee754-buffer) [![Docs](https://img.shields.io/badge/docs-online-blue.svg?style=for-the-badge)](https://rochars.github.io/ieee754-buffer/docs/index.html) [![Tests](https://img.shields.io/badge/tests-online-blue.svg?style=for-the-badge)](https://rochars.github.io/ieee754-buffer/test/dist/browser.html)  
[![Codecov](https://img.shields.io/codecov/c/github/rochars/ieee754-buffer.svg?style=flat-square)](https://codecov.io/gh/rochars/ieee754-buffer) [![Unix Build](https://img.shields.io/travis/rochars/ieee754-buffer.svg?style=flat-square)](https://travis-ci.org/rochars/ieee754-buffer) [![Windows Build](https://img.shields.io/appveyor/ci/rochars/ieee754-buffer.svg?style=flat-square&logo=appveyor)](https://ci.appveyor.com/project/rochars/ieee754-buffer) [![Scrutinizer](https://img.shields.io/scrutinizer/g/rochars/ieee754-buffer.svg?style=flat-square&logo=scrutinizer)](https://scrutinizer-ci.com/g/rochars/ieee754-buffer/)

**ieee754-buffer** is a module to encode and decode IEEE 754 floating point numbers.

- **MIT licensed**
- **Can be used where typed arrays can't**
- **Parse 16-bit, 32-bit and 64-bit floats**

## Install
```
npm install ieee754-buffer
```

## Use

### Browser
Use the **ieee754-buffer.js** file in the **/dist** folder:
```html
<script src="ieee754-buffer.js"></script>
<script>
	let packer = new IEEE754Buffer.IEEE754Buffer(8, 23);
	var buffer = [];
	packer.pack(buffer, 0, 3.1415927410);
	console.log(buffer);
</script>
```

Or load it from the [jsDelivr](https://cdn.jsdelivr.net/npm/ieee754-buffer) CDN:
```html
<script src="https://cdn.jsdelivr.net/npm/ieee754-buffer"></script>
```

Or load it from [unpkg](https://unpkg.com/ieee754-buffer):
```html
<script src="https://unpkg.com/ieee754-buffer"></script>
```

### Node
```javascript
const IEEE754Buffer = require('ieee754-buffer').IEEE754Buffer;
let packer = new IEEE754Buffer(8, 23);
let buffer = Uint8Array(4);
packer.pack(buffer, 0, 3.1415927410);
console.log(buffer);
```

## About
- **Overflows** are rounded towards **Infinity** and **-Infinity**.
- **NaN** is packed as quiet NaN. Both **quiet NaN** and **signaling NaN** can be unpacked.
- Support packing and unpacking **zero** and **negative zero**.
- Support packing and unpacking **Infinity** and **negative Infinity**
- **always little endian**, use [byte-data](https://www.github.com/rochars/byte-data) for a endianness-independent interface of this lib

## API
```javascript

export class IEEE754Buffer {

  /**
   * Pack a IEEE 754 floating point number.
   * @param {number} ebits The exponent bits.
   * @param {number} fbits The fraction bits.
   */
  constructor(ebits, fbits) {}

  /**
   * Pack a IEEE 754 floating point number.
   * @param {!Uint8Array|!Array<number>} buffer The buffer.
   * @param {number} num The number.
   * @param {number} index The index to write on the buffer.
   * @return {number} The next index to write on the buffer.
   */
  pack(buffer, index, num) {}

  /**
   * Unpack a IEEE 754 floating point number.
   * Derived from IEEE754 by DeNA Co., Ltd., MIT License. 
   * Adapted to handle NaN. Should port the solution to the original repo.
   * @param {!Uint8Array|!Array<number>} buffer The buffer.
   * @param {number} index The index to read from the buffer.
   * @return {number} The floating point number.
   */
  unpack(buffer, index) {}
}
```

## Contributing
**ieee754-buffer** welcomes all contributions from anyone willing to work in good faith with other contributors and the community. No contribution is too small and all contributions are valued.

See [CONTRIBUTING.md](https://github.com/rochars/ieee754-buffer/blob/master/CONTRIBUTING.md) for details.

### Style guide
**ieee754-buffer** code should follow the Google JavaScript Style Guide:  
https://google.github.io/styleguide/jsguide.html

### Code of conduct
This project is bound by a code of conduct: The [Contributor Covenant, version 1.4](https://github.com/rochars/ieee754-buffer/blob/master/CODE_OF_CONDUCT.md), also available at https://www.contributor-covenant.org/version/1/4/code-of-conduct.html

## LICENSE
Copyright (c) 2018-2019 Rafael da Silva Rocha.  
Copyright (c) 2013 DeNA Co., Ltd.  
Copyright (c) 2010, Linden Research, Inc

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# CHANGELOG

## 2.0.0 - 2019-12-02
- Fix: Throw TypeError if input is not a number.

## 1.0.1 - 2019-12-31
- Fix: rounding byte values must work the same on all browsers

## 1.0.0 - 2019-12-30
- New package structure:
	* dist file is "./dist/ieee754-buffer.js", a UMD served as "main"
	* ES6 source is "./index.js", served as "module"

## 0.2.1 (2018-08-06)
- Fix docstring in packFloatBits_()

## 0.2.0 (2018-08-06)
- API change: IEEE754Buffer.pack(buffer, num, index) instead of IEEE754Buffer.pack(buffer, index, num)

## 0.1.0 (2018-08-05)
- API change: IEEE754Buffer as a class

## 0.0.2 (2018-08-04)
- UMD compatible with IE6+


test/loader.js
/**
 * Copyright (c) 2018-2019 Rafael da Silva Rocha.
 * https://github.com/rochars/ieee754-buffer
 *
 */

let ieee754Buffer;

// UMD bundle
if (process.argv[3] == '--umd') {
	console.log('umd tests');
	ieee754Buffer = require('../dist/ieee754-buffer.js');

// Source
} else {
	require = require("esm")(module);
	global.module = module;
	console.log('Source tests');
	ieee754Buffer = require('../index.js');
}

module.exports = ieee754Buffer;


test/binary64-es3.js
/*
 * Copyright (c) 2018 Rafael da Silva Rocha.
 */

/**
 * @fileoverview Test IEEE-754 binary64 numbers.
 * @see https://github.com/rochars/ieee754-buffer
 * @see https://en.wikipedia.org/wiki/Double-precision_floating-point_format
 */
if (typeof window !=== 'undefined') {
var ieee = ieee754Buffer;
var assert = assert || require('assert');

describe('Binary64 numbers', function() {     
    // Zeros
    it('pack 0', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 0, 11, 52);
        assert.deepEqual(
        	buffer, 
            [0,0,0,0,0,0,0,0]);
    });
    
    it('unpack 0', function() {
        assert.equal(
        	ieee.unpack([0,0,0,0,0,0,0,0], 0, 11, 52), 
            0);
    });
    
    it('pack -0', function() {
        var buffer = [];
        ieee.pack(buffer, 0, -0, 11, 52);
        assert.deepEqual(
        	buffer, 
            [0,0,0,0,0,0,0,128]);
    });
    it('unpack -0', function() {
        assert.equal(
        	ieee.unpack([0,0,0,0,0,0,0,128], 0, 11, 52), 
            -0);
    });
    
    // NaN
    it('pack NaN', function() {
        var buffer = [];
        ieee.pack(buffer, 0, NaN, 11, 52);
        assert.deepEqual(
            buffer, 
            [0x00,0x00,0x00,0x00,0x00,0x00,0xf8,0x7f]);
    });
    it('unpack NaN (quiet)', function() {
        assert.ok(isNaN(ieee.unpack(
            [0x00,0x00,0x00,0x00,0x00,0x00,0xf8,0x7f], 0, 11, 52)));
    });
    it('unpack NaN (signaling)', function() {
        assert.ok(isNaN(ieee.unpack(
            [0xff,0xff,0xff,0xff,0xff,0xff,0xff,0x7f], 0, 11, 52)));
    });

    // Infinity
    it('pack Infinity', function() {
        // Python struct.pack('d', math.inf)
        var buffer = [];
        ieee.pack(buffer, 0, Infinity, 11, 52);
        assert.deepEqual(
        	buffer, 
            [0x00,0x00,0x00,0x00,0x00,0x00,0xf0,0x7f]);
    });
    it('unpack Infinity', function() {
        assert.equal(
        	ieee.unpack([0x00,0x00,0x00,0x00,0x00,0x00,0xf0,0x7f], 0, 11, 52), 
            Infinity);
    });
    it('pack -Infinity', function() {
        // Python struct.pack('d', -math.inf)
        var buffer = [];
        ieee.pack(buffer, 0, -Infinity, 11, 52);
        assert.deepEqual(
        	buffer, 
            [0x00,0x00,0x00,0x00,0x00,0x00,0xf0,0xff]);
    });
    it('unpack -Infinity', function() {
        assert.equal(
        	ieee.unpack([0x00,0x00,0x00,0x00,0x00,0x00,0xf0,0xff], 0, 11, 52), 
            -Infinity);
    });
    
    // Pi
    it('pack pi as 3.1415926535897931', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 3.1415926535897931, 11, 52);
        assert.deepEqual(
            buffer, 
            [0x18,0x2d,0x44,0x54,0xfb,0x21,0x09,0x40]);
    });
    it('unpack pi as 3.1415926535897931', function() {
        assert.equal(
            ieee.unpack([0x18,0x2d,0x44,0x54,0xfb,0x21,0x09,0x40], 0, 11, 52).toFixed(15), 
            '3.141592653589793');
    });
    
    // struct.unpack('d', b'\xd9\x7d\xda\xf5\xd0\xf2\xbe\x3a')
    it('pack 1e-25', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 1e-25, 11, 52);
        assert.deepEqual(
            buffer,
            [0xd9,0x7d,0xda,0xf5,0xd0,0xf2,0xbe,0x3a]);
    });
    it('unpack 1e-25', function() {
        assert.equal(
            ieee.unpack([0xd9,0x7d,0xda,0xf5,0xd0,0xf2,0xbe,0x3a], 0, 11, 52),
            1e-25);
    });

    // Random values
    it('pack -1', function() {
        var buffer = [];
        ieee.pack(buffer, 0, -1, 11, 52);
        assert.deepEqual(
            buffer,
            [0,0,0,0,0,0,240,191]);
    });
    it('pack -0.5', function() {
        var buffer = [];
        ieee.pack(buffer, 0, -0.5, 11, 52, 16);
        assert.deepEqual(
            buffer,
            [0,0,0,0,0,0,224,191]);
    });
    it('pack 0.5', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 0.5, 11, 52);
        assert.deepEqual(
            buffer,
            [0,0,0,0,0,0,224,63]);
    });
    it('pack 9', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 9, 11, 52);
        assert.deepEqual(
            buffer,
            [0,0,0,0,0,0,34,64]);
    });
    it('pack 31.41592653589793', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 31.41592653589793, 11, 52);
        assert.deepEqual(
            buffer,
            [94,56,85,41,122,106,63,64]);
    });
    it('pack 314159265358979.3', function() {
        // struct.pack() == b'\x35\x48\xa2\x76\x9e\xdb\xf1\x42'
        var buffer = [];
        ieee.pack(buffer, 0, 314159265358979.3, 11, 52);
        assert.deepEqual(
            buffer,
            //[53,72,162,118,158,219,241,66]);
            [0x35,0x48,0xa2,0x76,0x9e,0xdb,0xf1,0x42]);
    });
    it('unpack 314159265358979.3', function() {
        assert.deepEqual(
            ieee.unpack([0x35,0x48,0xa2,0x76,0x9e,0xdb,0xf1,0x42], 0, 11, 52),
            314159265358979.3);
    });
    it('pack 0', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 0, 11, 52);
        assert.deepEqual(
            buffer,
            [0,0,0,0,0,0,0,0]);
    });
    it('pack 2', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 2, 11, 52);
        assert.deepEqual(
            buffer,
            [0,0,0,0,0,0,0,64]);
    });
});
}

test/src/binary16.js
/*
 * Copyright (c) 2018 Rafael da Silva Rocha.
 */

/**
 * @fileoverview Test IEEE-754 binary16 numbers.
 * @see https://github.com/rochars/ieee754-buffer
 * @see https://en.wikipedia.org/wiki/Half-precision_floating-point_format
 */

var ieee754 = require('../../test/loader.js');
var assert = assert || require('assert');

describe('Binary16 numbers', function() {     

    let ieee = new ieee754.IEEE754Buffer(5, 11);

    // Zeros
    it('pack 0', function() {
        var buffer = [];
        ieee.pack(buffer, 0, 0);
        assert.deepEqual(
        	buffer, 
            [0,0]);
    });
    
    it('unpack 0', function() {
        assert.equal(
        	ieee.unpack([0,0], 0), 
            0);
    });
    it('pack -0', function() {
        var buffer = [];
        ieee.pack(buffer, -0, 0);
        assert.deepEqual(
        	buffer, 
            [0,128]);
    });
    it('unpack -0', function() {
        assert.equal(
        	ieee.unpack([0,128],  0), 
            -0);
    });
    it('pack 1e-25', function() {
        var buffer = [];
        ieee.pack(buffer, 1e-25, 0);
        assert.deepEqual(
            buffer, 
            [0,0]);
    });
    it('unpack 0', function() {
        assert.equal(
            ieee.unpack([0,0],  0), 
            0);
    });

    // NaN
    it('pack NaN', function() {
        var buffer = [];
        ieee.pack(buffer, NaN, 0);
        assert.deepEqual(
            buffer, 
            [0, 126]); // Python struct.pack('e', math.nan)
    });
    it('unpack NaN', function() {
        assert.ok(isNaN(ieee.unpack([0, 126],  0)));
    });

    // Infinity
    it('pack Infinity', function() {
        var buffer = [];
        ieee.pack(buffer, Infinity, 0);
        assert.deepEqual(
        	buffer, 
            [0, 124]); // Python struct.pack('e', math.inf)
    });
    it('unpack Infinity', function() {
        assert.equal(
        	ieee.unpack([0, 124],  0), 
            Infinity);
    });
    it('pack -Infinity', function() {
        var buffer = [];
        ieee.pack(buffer, -Infinity, 0);
        assert.deepEqual(
        	buffer, 
            [0x00, 0xfc]); // Python struct.pack('e', -math.inf)
    });
    it('unpack -Infinity', function() {
        assert.equal(
        	ieee.unpack([0x00, 0xfc],  0), 
            -Infinity);
    });
    // round to Infinity
    it('round 65520 to Infinity when packing', function() {
        var buffer = [];
        ieee.pack(buffer, 65520, 0);
        assert.deepEqual(
            buffer, 
            [0, 124]);
    });
    it('round 65520+ to Infinity when packing', function() {
        var buffer = [];
        ieee.pack(buffer, 65521, 0);
        assert.deepEqual(
            buffer, 
            [0, 124]);
    });

    // Pi
    it('pack pi as 3.14159', function() {
        var buffer = [];
        ieee.pack(buffer, 3.14159, 0);
        assert.deepEqual(buffer, 
            [0x48,0x42]); // Python struct.pack('e', 3.14159)
    });
    it('unpack pi as 3.140625', function() {
        // Use toFixed(6) here to avoid JS rounding differences with Python.
        // Python struct.pack is used to check the data in most of the tests
        assert.equal(ieee.unpack([0x48,0x42],  0).toFixed(6), 
            '3.140625');
    });
    
    // Exact representations
    // Test 1, MIN, MAX
    it('pack 1', function() {
        var buffer = [];
        ieee.pack(buffer, 1, 0);
        assert.deepEqual(
        	buffer, 
            [0x00,0x3c]);
    });
    it('unpack 1', function() {
        assert.equal(
        	ieee.unpack([0x00,0x3c],  0), 
            1);
    });
    it('pack -2048 (min exact)', function() {
        var buffer = [];
        ieee.pack(buffer, -2048, 0);
        assert.deepEqual(
        	buffer, 
            [0x00,0xe8]);
    });
    it('unpack -2048 (min exact)', function() {
        assert.equal(
        	ieee.unpack([0x00,0xe8],  0), 
            -2048);
    });
    it('pack 2048 (max exact)', function() {
        var buffer = [];
        ieee.pack(buffer, 2048, 0);
        assert.deepEqual(
        	buffer, 
            [0x00,0x68]);
    });
    it('unpack 2048 (max exact)', function() {
        assert.equal(
        	ieee.unpack([0x00,0x68],  0), 
            2048);
    });
    
    // Rounding
    it('pack 2049 like it pack 2048', function() {
        var buffer = [];
        ieee.pack(buffer, 2049, 0);
        assert.deepEqual(
            buffer, 
            [0x00,0x68]);
    });
    it('pack 2050', function() {
        var buffer = [];
        ieee.pack(buffer, 2050, 0);
        assert.deepEqual(
            buffer, 
            [0x01,0x68]);
    });
    it('unpack 2050', function() {
        assert.deepEqual(
            ieee.unpack([0x01,0x68],  0), 
            2050);
    });
    it('pack 2051 like it pack 2050', function() {
        var buffer = [];
        ieee.pack(buffer, 2051, 0);
        assert.deepEqual(
            buffer, 
            [0x01,0x68]);
    });
});


rollup.config.js
/*
 * Copyright (c) 2018-2019 Rafael da Silva Rocha.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

/**
 * @fileoverview rollup configuration file.
 * @see https://github.com/rochars/ieee754-buffer
 */

import compiler from '@ampproject/rollup-plugin-closure-compiler';

export default [
  {
    input: 'index.js',
    output: [
      {
        file: 'dist/ieee754-buffer.js',
        name: 'IEEE754Buffer',
        format: 'umd'
      },
    ],
    plugins: [
      compiler({
        language_in: 'ECMASCRIPT6',
        language_out: 'ECMASCRIPT3',
        compilation_level: 'SIMPLE',
        warning_level: 'VERBOSE'
      })
    ]
  }
];


ieee754-buffer.d.ts
// Type definitions for ieee754-buffer 0.1.0
// Project: https://github.com/rochars/ieee754-buffer
// Definitions by: Rafael da Silva Rocha <https://github.com/rochars>
// Definitions: https://github.com/rochars/ieee754-buffer

export {IEEE754Buffer};

declare class IEEE754Buffer {

	/**
	 * Pack a IEEE 754 floating point number.
	 * @param {number} ebits The exponent bits.
	 * @param {number} fbits The fraction bits.
	 */
	constructor(ebits, fbits);

	/**
	 * Pack a IEEE 754 floating point number.
	 * @param {!Uint8Array|!Array<number>} buffer The buffer.
	 * @param {number} index The index to write on the buffer.
	 * @param {number} num The number.
	 * @return {number} The next index to write on the buffer.
	 */
	pack(
		buffer: Uint8Array|number[],
		index: number,
		num: number): number;

	/**
	 * Unpack a IEEE 754 floating point number.
	 * @param {!Uint8Array|!Array<number>} buffer The buffer.
	 * @param {number} index The index to read from the buffer.
	 * @return {number} The floating point number.
	 */
	unpack(
		buffer: Uint8Array|number[],
		index: number): number;
}


externs/ieee754-buffer.js
/*
 * Copyright (c) 2018-2019 Rafael da Silva Rocha.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

/**
 * @fileoverview Externs for uint-buffer 1.0
 * @see https://github.com/rochars/uint-buffer
 * @externs
 */

/**
 * A class to write and read unsigned ints to and from byte buffers.
 */
var IEEE754Buffer = {};

/**
 * Pack a IEEE 754 floating point number.
 * @param {!Uint8Array|!Array<number>} buffer The buffer.
 * @param {number} num The number.
 * @param {number} index The index to write on the buffer.
 * @return {number} The next index to write on the buffer.
 */
IEEE754Buffer.prototype.pack = function(buffer, num, index=0) {};

/**
 * Unpack a IEEE 754 floating point number.
 * @param {!Uint8Array|!Array<number>} buffer The buffer.
 * @param {number} index The index to read from the buffer.
 * @return {number} The floating point number.
 */
IEEE754Buffer.prototype.unpack = function(buffer, index=0) {};


ieee754-buffer.umd.js
;(function (global, factory) {typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :typeof define === 'function' && define.amd ? define(factory) :(global.IEEE754Buffer = factory());}(this, (function () {
var IEEE754Buffer = (function (exports) {
  var classCallCheck = function (instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  };

  var createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  /*
   * Copyright (c) 2018 Rafael da Silva Rocha.
   * Copyright (c) 2013 DeNA Co., Ltd.
   * Copyright (c) 2010, Linden Research, Inc
   *
   * Permission is hereby granted, free of charge, to any person obtaining
   * a copy of this software and associated documentation files (the
   * "Software"), to deal in the Software without restriction, including
   * without limitation the rights to use, copy, modify, merge, publish,
   * distribute, sublicense, and/or sell copies of the Software, and to
   * permit persons to whom the Software is furnished to do so, subject to
   * the following conditions:
   *
   * The above copyright notice and this permission notice shall be
   * included in all copies or substantial portions of the Software.
   *
   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
   * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
   * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
   * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
   * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
   * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
   *
   */

  /**
   * @fileoverview Encode and decode IEEE 754 floating point numbers.
   * @see https://github.com/rochars/ieee754-buffer
   * @see https://bitbucket.org/lindenlab/llsd/raw/7d2646cd3f9b4c806e73aebc4b32bd81e4047fdc/js/typedarray.js
   * @see https://github.com/kazuho/ieee754.js/blob/master/ieee754.js
   */

  /** 
   * @module IEEE754Buffer
   * @ignore
   */

  /**
   * A class to encode and decode IEEE 754 floating-point numbers.
   */
  var IEEE754Buffer = function () {

    /**
     * Pack a IEEE 754 floating point number.
     * @param {number} ebits The exponent bits.
     * @param {number} fbits The fraction bits.
     */
    function IEEE754Buffer(ebits, fbits) {
      classCallCheck(this, IEEE754Buffer);

      this.ebits = ebits;
      this.fbits = fbits;
      this.bias = (1 << ebits - 1) - 1;
      this.numBytes = Math.ceil((ebits + fbits) / 8);
      this.biasP2 = Math.pow(2, this.bias + 1);
      this.ebitsFbits = ebits + fbits;
      this.fbias = Math.pow(2, -(8 * this.numBytes - 1 - ebits));
    }

    /**
     * Pack a IEEE 754 floating point number.
     * @param {!Uint8Array|!Array<number>} buffer The buffer.
     * @param {number} num The number.
     * @param {number} index The index to write on the buffer.
     * @return {number} The next index to write on the buffer.
     */


    createClass(IEEE754Buffer, [{
      key: "pack",
      value: function pack(buffer, num, index) {
        // Round overflows
        if (Math.abs(num) > this.biasP2 - this.ebitsFbits * 2) {
          num = num < 0 ? -Infinity : Infinity;
        }
        /**
         * sign, need this to handle negative zero
         * @see http://cwestblog.com/2014/02/25/javascript-testing-for-negative-zero/
         * @type {number}
         */
        var sign = ((num = +num) || 1 / num) < 0 ? 1 : num < 0 ? 1 : 0;
        num = Math.abs(num);
        /** @type {number} */
        var exp = Math.min(Math.floor(Math.log(num) / Math.LN2), 1023);
        /** @type {number} */
        var fraction = this.roundToEven(num / Math.pow(2, exp) * Math.pow(2, this.fbits));
        // NaN
        if (num !== num) {
          fraction = Math.pow(2, this.fbits - 1);
          exp = (1 << this.ebits) - 1;
          // Number
        } else if (num !== 0) {
          if (num >= Math.pow(2, 1 - this.bias)) {
            if (fraction / Math.pow(2, this.fbits) >= 2) {
              exp = exp + 1;
              fraction = 1;
            }
            // Overflow
            if (exp > this.bias) {
              exp = (1 << this.ebits) - 1;
              fraction = 0;
            } else {
              exp = exp + this.bias;
              fraction = this.roundToEven(fraction) - Math.pow(2, this.fbits);
            }
          } else {
            fraction = this.roundToEven(num / Math.pow(2, 1 - this.bias - this.fbits));
            exp = 0;
          }
        }
        return this.packFloatBits_(buffer, index, sign, exp, fraction);
      }

      /**
       * Unpack a IEEE 754 floating point number.
       * Derived from IEEE754 by DeNA Co., Ltd., MIT License. 
       * Adapted to handle NaN. Should port the solution to the original repo.
       * @param {!Uint8Array|!Array<number>} buffer The buffer.
       * @param {number} index The index to read from the buffer.
       * @return {number} The floating point number.
       */

    }, {
      key: "unpack",
      value: function unpack(buffer, index) {
        /** @type {number} */
        var eMax = (1 << this.ebits) - 1;
        /** @type {number} */
        var significand = void 0;
        /** @type {string} */
        var leftBits = "";
        for (var i = this.numBytes - 1; i >= 0; i--) {
          /** @type {string} */
          var t = buffer[i + index].toString(2);
          leftBits += "00000000".substring(t.length) + t;
        }
        /** @type {number} */
        var sign = leftBits.charAt(0) == "1" ? -1 : 1;
        leftBits = leftBits.substring(1);
        /** @type {number} */
        var exponent = parseInt(leftBits.substring(0, this.ebits), 2);
        leftBits = leftBits.substring(this.ebits);
        if (exponent == eMax) {
          if (parseInt(leftBits, 2) !== 0) {
            return NaN;
          }
          return sign * Infinity;
        } else if (exponent === 0) {
          exponent += 1;
          significand = parseInt(leftBits, 2);
        } else {
          significand = parseInt("1" + leftBits, 2);
        }
        return sign * significand * this.fbias * Math.pow(2, exponent - this.bias);
      }

      /**
       * Pack a IEEE754 from its sign, exponent and fraction bits
       * and place it in a byte buffer.
       * @param {!Uint8Array|!Array<number>} buffer The byte buffer to write to.
       * @param {number} index The buffer index to write.
       * @param {number} ebits The number of bits of the exponent.
       * @param {number} fbits The number of bits of the fraction.
       * @param {number} sign The sign.
       * @param {number} exp the exponent.
       * @param {number} fraction The fraction.
       * @return {number}
       * @private
       */

    }, {
      key: "packFloatBits_",
      value: function packFloatBits_(buffer, index, sign, exp, fraction) {
        /** @type {!Array<number>} */
        var bits = [];
        // the sign
        bits.push(sign);
        // the exponent
        for (var i = this.ebits; i > 0; i -= 1) {
          bits[i] = exp % 2 ? 1 : 0;
          exp = Math.floor(exp / 2);
        }
        // the fraction
        var len = bits.length;
        for (var _i = this.fbits; _i > 0; _i -= 1) {
          bits[len + _i] = fraction % 2 ? 1 : 0;
          fraction = Math.floor(fraction / 2);
        }
        // pack as bytes
        /** @type {string} */
        var str = bits.join('');
        /** @type {number} */
        var numBytes = this.numBytes + index - 1;
        /** @type {number} */
        var k = index;
        while (numBytes >= index) {
          buffer[numBytes] = parseInt(str.substring(0, 8), 2);
          str = str.substring(8);
          numBytes--;
          k++;
        }
        return k;
      }
    }, {
      key: "roundToEven",
      value: function roundToEven(n) {
        var w = Math.floor(n),
            f = n - w;
        if (f < 0.5) {
          return w;
        }
        if (f > 0.5) {
          return w + 1;
        }
        return w % 2 ? w + 1 : w;
      }
    }]);
    return IEEE754Buffer;
  }();

  exports.IEEE754Buffer = IEEE754Buffer;

  return exports;

}({}));
return IEEE754Buffer;})));
